diagnostic_rt_qpcr/operation_types/run_qpcr/protocol.rb

```diff
@@ -62,38 +76,148 @@ def main
+
+      op.associate(RAW_QPCR_DATA_KEY, file_name)
+
+      thermocycler = ThermocyclerFactory.build(
+        model: thermo_type['model']
+      )
+
+      go_to_thermocycler(thermocycler_name: thermo_type['name'], plate: plate)
+
+      set_up_program(
+        thermocycler: thermocycler,
+        program: program,
+        composition: composition,
+        qpcr: @job_params[:qpcr]
+      )
+
+      load_plate_and_start_run(
+        thermocycler: thermocycler,
+        items: plate,
+        experiment_filename: file_name
+      )
+      running_thermocyclers.push([op, thermocycler])
+    end
+    running_thermocyclers
+  end
+
+  def associate_measurement(file_name:, plate:)
+    file = upload_data(file_name, 1, 4)
+    plate.associate(RAW_QPCR_DATA_KEY, file)
+  end
+
+  def go_to_thermocycler(thermocycler_name:, plate: nil)
+    show do
+      title 'Go to Thermocycler'
+      note "Take the #{plate.object_type.name} <b>#{plate.id}</b>"\
+           " to Thermocycler #{thermocycler_name}" unless plate.nil?
+      note "Complete the next few steps at Thermocycler #{thermocycler_name}"
+    end
+  end
+
+  def get_available_thermocyclers
+    thermocyclers = find_thermocyclers
+    available_key = 'available'
+    response = show do
+      title 'Check Available Thermocyclers'
+      note 'Please check which thermocyclers are currently available'
+      thermocyclers.each_with_index do |thermo|
+        select([available_key, 'unavailable'],
+               var: thermo['name'].to_s,
+               label: "Thermocycler #{thermo['name']}")
+      end
+    end
+    available_thermo = []
+    thermocyclers.map do |thermo|
+      next unless response[thermo['name'].to_s] == available_key || debug
+
+      available_thermo.push(thermo)
+    end
+    available_thermo
+  end
+
+  def find_thermocyclers
+    Parameter.where(key: 'thermocycler').map { |thr| JSON.parse(thr.value) }
+  end
+
+  def remove_unavailable_operations(available_thermocyclers)
+    ops_to_remove = find_unavailable_ops(available_thermocyclers)
+    operations.reject! { |op| ops_to_remove.include?(op) }
+    error_op_warning(ops_to_remove)
+    ops_to_remove.each do |op|
+      op.error(:unavailablethermocycler, 'No thermocyclers were available')
+      op.set_status_recursively('pending')
+    end
+  end
+
+  def error_op_warning(ops_to_remove)
+    show do
+      title 'Thermocyclers Unavailable'
+      note 'There are not enough available thermocyclers for this job'
+      warning 'The following plates were removed from this job'
+      ops_to_remove.each do |op|
+        note op.input(INPUT_REACTIONS).collection.id.to_s
+      end
+    end
+  end
+
+  def find_unavailable_ops(thermocyclers)
+    ops_to_remove = []
+    operations.each do |op|
+      available = false
+      thermocyclers.each do |thermo|
+        next unless thermo['model'] == op.temporary[:options][:thermocycler_model] || true
+
+        op.associate(THERMOCYCLER_KEY, thermo)
```

Comment:
This actually seems to be the critical line in this series of methods. I think it would be cleaner to just make this association until you run out of thermocyclers or operations, and then select the operations without this association to be sent back to pending. Also, `operations.zip(thermocyclers)` (https://ruby-doc.org/core-2.7.0/Array.html#method-i-zip) would be a cleaner way to pair them up.

Reply:
I previously had `zip` however I need to make sure that the thermocycler that it is matched with is the type that was specified in the `options` param.  Can zip do this?